### PR TITLE
[expr.prim.lambda] Split specification of lambda expressions into sub…

### LINKED ISSUE #1155: [expr.prim.lambda] p 6, 7 obnoxiously long
Paragraphs 6 and 7 of [expr.prim.lambda] are very long. P 7 contains 2 notes and 3 examples!

Could we please rewrite those to become multiple paragraphs? Consider having long examples be paragraphs on their own.

@jensmaurer, @zygoloid, any takers?

### PR DESCRIPTION
**Pull request by jensmaurer**
…sections.

Fixes #1155.

### CODE REVIEW

**Comment by jensmaurer** [comment]
I have tried to restrain myself to reshuffling text and introducing subsection headings.

File: source/expressions.tex [hunk 1]
```diff
@@ -706,14 +706,35 @@
...
-This class type is not an aggregate type~(\ref{dcl.init.aggr}).
+whose properties are described in section~\ref{expr.prim.lambda.closure}.
+
+\pnum
+If a \grammarterm{lambda-expression} does not include a
+\grammarterm{lambda-declarator}, it is as if the \grammarterm{lambda-declarator} were
+\tcode{()}.
+The lambda return type is \tcode{auto}, which is replaced by the
+type specified by the
+\grammarterm{trailing-return-type} if provided and/or deduced from
```
Comment (tkoeppe) [inline source/expressions.tex]: How about just "or" instead of "and/or"?

Comment (jensmaurer) [inline source/expressions.tex]: This is unchanged text, just moved around. Not my invention.

File: source/expressions.tex [hunk 2]
```diff
@@ -788,7 +792,9 @@
 q();                                                   // OK: outputs \tcode{1a3.14}
 \end{codeblock}
 \end{example}
-This function call operator or operator template is declared
+
```
Comment (tkoeppe) [inline source/expressions.tex]: (I cannot comment on unchanged code, so this comment is out of place.)

The example seems to be specific to packs only, not to the entire paragraph. What do you think of the following: In lines old-762/new-766, take the sentence "The invented type ..." out and end the paragraph after the next sentence, just before the example, Then put the removed sentence back at the start of the now new paragraph, followed by the example. That way the new paragraph concentrates on one specific detail (packs), together with example.

Comment (jensmaurer) [inline source/expressions.tex]: First, the example also has non-pack things, e.g. glambda at the top.
Second, "The invented type template-parameter ..." refers to the template-parameter introduced in the preceding sentence. Moving this to a separate paragraph, with something else in between, would lose the connection.

Comment (tkoeppe) [inline source/expressions.tex]: OK, the example is a good point. I wouldn't find losing the connection too worrying in general; I think it's acceptable for later things to refer to earlier things. That's more of a judgement call. But we can keep it as is.

Comment (jensmaurer) [inline source/expressions.tex]: Yes, but if you read the intermediate sentence, it goes off on a tangent, and returning to the template-parameter later is odd.

**Comment by jensmaurer** [comment]
Taking no action on the review comments, as explained above.

@zygoloid: You're next.

**Comment by jensmaurer** [comment]
I've updated some cross-references to now refer to a more specific subsection, where appropriate.

**Comment by jensmaurer** [comment]
No-change rebase.

**Comment by zygoloid** [comment]
1. [expr.prim.lambda]p2 seems out of place, and breaks the natural flow from p1 to p3. Maybe we can move it down next to where we talk about how these *decl-specifier*s map into properties of the closure type's `operator()`? Or if not there, then at least to after p5.

2. I think [expr.prim.lambda]p4 would be better as the first paragraph of [expr.prim.lambda.closure], as it's a good summary of what we're about to describe in there, slightly tweaked to something like:

> The type of a *lambda-expression* (which is also the type of the closure object) is a unique, unnamed non-union class type, called the *closure type*, whose properties are described below.

3. The start of [expr.prim.lambda.capture] is quite jarring. We could do with an introductory sentence explaining what on earth a capture is before we get into the nitty-gritty details. Maybe something like:

> The body of a *lambda-expression* may refer to automatic storage duration variables and the `*this` object of enclosing block scopes by capturing those entities, as described below.

4. We might also want to consider moving the grammar description of explicit captures down into the captures subclause.

**Comment by jensmaurer** [comment]
re 1: Moved normative sentence from [expr.prim.lambda]p2 to the start of [expr.prim.lambda.closure]p3. Moved example as a new paragraph to after [expr.prim.lambda.closure]p3.

re 2: Done.

re 3: Done.  I'm not super-happy with the "*this object". I added "(if any)", since we might not be in a member function.

re 4: Why just the explicit captures? Seems like all of the capture grammar could go there. Done.

**Comment by jensmaurer** [comment]
Subject: 	Re: editorial restructuring of [class.copy]
Date: 	Tue, 31 Jan 2017 14:56:02 -0500
From: 	William M. (Mike) Miller <william.m.miller@gmail.com>

[...]
The ones for expr.prim.lambda are involved enough that I think I'd like to have CWG take a look.

**Comment by jensmaurer** [comment]
Uploaded to CWG wiki for Kona, as agreed with Mike Miller:
[expr.prim.lambda](http://wiki.edg.com/pub/Wg21kona2017/CoreWorkingGroup/expr-prim-lambda.pdf)

**Comment by jensmaurer** [comment]
CWG suggested to move the first sentence of closure types p4 to the general section and otherwise approved this change.

**Comment by jensmaurer** [comment]
Rebased and implemented CWG guidance.

**Comment by tkoeppe** [comment]
@jensmaurer: Rebase, just to be sure?

**Pull request merged**